substrate/client/network/statement/src/lib.rs

```diff
@@ -442,55 +517,620 @@ where
 	}
 
 	/// Propagate one statement.
-	pub fn propagate_statement(&mut self, hash: &Hash) {
+	pub async fn propagate_statement(&mut self, hash: &Hash) {
 		// Accept statements only when node is not major syncing
 		if self.sync.is_major_syncing() {
 			return
 		}
 
 		log::debug!(target: LOG_TARGET, "Propagating statement [{:?}]", hash);
 		if let Ok(Some(statement)) = self.statement_store.statement(hash) {
-			self.do_propagate_statements(&[(*hash, statement)]);
+			self.do_propagate_statements(&[(*hash, statement)]).await;
 		}
 	}
 
-	fn do_propagate_statements(&mut self, statements: &[(Hash, Statement)]) {
-		let mut propagated_statements = 0;
-
+	async fn do_propagate_statements(&mut self, statements: &[(Hash, Statement)]) {
+		log::debug!(target: LOG_TARGET, "Propagating {} statements for {} peers", statements.len(), self.peers.len());
 		for (who, peer) in self.peers.iter_mut() {
+			log::trace!(target: LOG_TARGET, "Start propagating statements for {}", who);
+
 			// never send statements to light nodes
 			if matches!(peer.role, ObservedRole::Light) {
+				log::trace!(target: LOG_TARGET, "{} is a light node, skipping propagation", who);
 				continue
 			}
 
 			let to_send = statements
 				.iter()
 				.filter_map(|(hash, stmt)| peer.known_statements.insert(*hash).then(|| stmt))
```

Comment:
What would happen if the `send_async_notification` is limited by the timeout here? Our view of the peer would be that the peer received the statement and we'll not retry later, should we only insert on success?